telemetry/src/ws.rs

```diff
@@ -18,58 +29,157 @@ pub async fn start(config: &Configuration, telemetry: Receiver<Telemetry>) -> Re
+        name: String,
+        sink: S,
+        sink_factory: F,
+        retry_period: RetryPeriod,
+        internal_sender: Sender<InternalMessage>,
+    ) -> Self {
+        Self {
+            name,
+            sink_factory,
+            retry_period,
+            internal_sender,
+            sink: Some(sink),
+            init_msg: None,
+        }
+    }
+
+    pub async fn run(
+        mut self,
+        receiver: Receiver<Telemetry>,
+        internal_receiver: Receiver<InternalMessage>,
+    ) {
+        let mut stream = ReceiverStream::new(receiver).fuse();
+        let mut internal_stream = ReceiverStream::new(internal_receiver).fuse();
+        #[allow(clippy::restriction)]
+        loop {
+            tokio::select! {
+                msg = stream.next() => {
+                    if let Some(msg) = msg {
+                        self.on_telemetry(msg).await;
+                    } else {
+                        break;
                     }
-                    Err(e) => {
-                        iroha_logger::error!("send failed: {:?}", e);
-                        // TBD: What is the proper way to signal about it?
+                }
+                msg = internal_stream.next() => {
+                    if let Some(InternalMessage::Reconnect) = msg {
+                        self.on_reconnect().await;
                     }
                 }
             }
+        }
+    }
+
+    async fn on_telemetry(&mut self, telemetry: Telemetry) {
+        match prepare_message(&self.name, telemetry) {
+            Ok((msg, msg_kind)) => {
+                if let Some(MessageKind::Initialization) = msg_kind {
+                    self.init_msg = Some(msg.clone());
+                }
+                self.send_msg(msg).await;
+            }
             Err(e) => {
                 iroha_logger::error!("prepare_message failed: {:?}", e);
-                // TBD: What is the proper way to signal about it?
             }
         }
     }
-}
 
-fn prepare_message(name: &str, telemetry: Telemetry) -> Result<Message> {
-    enum Msg {
-        SystemConnected,
-        Other,
+    async fn on_reconnect(&mut self) {
+        if let Ok(sink) = self.sink_factory.create().await {
+            if let Some(msg) = self.init_msg.as_ref() {
+                iroha_logger::debug!("Reconnected telemetry");
+                self.sink = Some(sink);
+                let msg = msg.clone();
+                self.send_msg(msg).await;
+            } else {
+                // The reconnect is required if sending a message fails.
+                // The first message to be sent is initialization.
+                // The path is assumed to be unreachable.
+                iroha_logger::error!(
+                    "Cannot reconnect telemetry because there is no initialization message"
+                );
+            }
+        } else {
+            self.schedule_reconnect();
+        }
     }
+
+    async fn send_msg(&mut self, msg: Message) {
```

Comment:
```suggestion
    async fn send_message(&mut self, msg: Message) {
```
As we have `prepare_message` function right after here